### PR TITLE
break `handlers.jl` into default_render, default_equality and preprocess

### PR DESCRIPTION
**Pull request by johnnychen94**
This PR is a part of #36 that makes the overall pipeline cleaner. It's also a preparation for the introduction of `@test_reference_broken` and `@test_reference_skip`.

To avoid code duplication, codes in `handler.jl` are broken down into three functions: 

* `default_render` that infers the most appropriate render mode according to types of `filename` and `actual`. By default, it's `BeforeAfterFull()`.
* `default_equality` that infers the suitable equality compare method. By default, it's `isequal`.
* `preprocess` that makes `actual` have the same type of `reference`.

### CODE REVIEW

File: src/core.jl [hunk 1]
```diff
@@ -141,8 +213,14 @@ function _test_reference(equiv, rendermode, file::File, actual::T) where T
...
 
     # file exists
+    actual = preprocess(file, actual; kw...)
     reference = loadfile(T, file)
 
+    if isnothing(equiv)
+        # generally, `reference` and `actual` are of the same type after preprocessing
+        equiv = default_equality(typeof(reference), typeof(actual))
+    end
+
```
Comment (johnnychen94) [inline src/core.jl]: infer the default equality method after preprocessing because now `reference` and `actual` have the same type.

File: src/test_reference.jl [hunk 2]
```diff
@@ -83,10 +83,6 @@ macro test_reference(reference, actual, kws...)
     expr
 end
 
-function test_reference(filename::AbstractString, actual; kw...)
-    test_reference(query_extended(filename), actual; kw...)
-end
-
```
Comment (johnnychen94) [inline src/test_reference.jl]: moved to `core.jl`

File: src/core.jl [hunk 3]
```diff
@@ -92,6 +92,25 @@ function render(mode::RenderMode, actual)
...
+default_rendermode(::TestMode, ::File, actual) = BeforeAfterFull()
+
+# plain TXT
+default_rendermode(::TestMode, ::File{format"TXT"}, actual) = BeforeAfterLimited()
+default_rendermode(::TestMode, ::File{format"TXT"}, actual::AbstractString) = Diff()
+default_rendermode(::TestMode, ::File{format"TXT"}, actual::Number) = BeforeAfterFull()
+
+# SHA256
+default_rendermode(::TestMode, ::File{format"SHA256"}, actual) = BeforeAfterFull()
```
Comment (johnnychen94) [inline src/core.jl]: This is redundant; I put it here to make sure `SHA256` is not influenced by potential future changes of the `::File` fallback.

File: src/core.jl [hunk 4]
```diff
@@ -118,13 +137,66 @@ function savefile(file::TextFile, content)
     write(file.filename, content)
 end
 
+#######################################
+# Preprocess
+
+preprocess(::File, actual; kw...) = actual
```
Comment (johnnychen94) [inline src/core.jl]: `kw...` here is to ensure compatibility to `size` keyword for image types

File: src/core.jl [hunk 5]
```diff
@@ -118,13 +137,66 @@ function savefile(file::TextFile, content)
...
+
+    test_reference(query_extended(filename), actual, render, by; kw...)
+end
+
+function test_reference(
+    file::File,
+    actual::T,
+    rendermode::Union{RenderMode, Nothing}=nothing,
+    equiv::Union{Function, Nothing}=nothing;
+    kw...) where T
```
Comment (johnnychen94) [inline src/core.jl]: Arguments order is changed because I made `rendermode` and `equiv` optional.

`kw...` are all passed to `preprocess` as it was (the only possible `kw` here is `size` for image type). It might be changed in the future, so I didn't document it.

**Comment by oxinabox** [comment]
What is the motivation of moving all this code out of its own file and intio `core.jl`?
Won't that make it harder to find things?

From the title I thought it was going to turn `handlers.jl` into 3 files:
`rendered.jl` `equality_measures.jl` and `preprocessors.jl`.


Aside: moving code and editting code makes for hard to follow diffs.

**Comment by johnnychen94** [comment]
> What is the motivation of moving all this code out of its own file and intio core.jl?

I'm planing to move `test_reference` and `_test_reference` from `core.jl` to `test_reference.jl`. After that, what are left in `core.jl` are other things related to the core pipeline.

**Comment by johnnychen94** [comment]
@oxinabox I've tried to minimize the differences for you to review. 

Since all codes in `handlers.jl` are removed and rewrite using `default_rendermode`, `default_equality` and `preprocess`. It's unavoidable hard to compare and review the diff. I've double-checked them and I'm confident about the logic here.

It comes to an end of my vacation, I need to focus on my school stuff first and then I'll come back continue it.

Since you prefer holding things apart in separate files, I'll move things related to `RenderMode` to `render.jl` when this PR is ready to be merged. 

After this PR:

* Documentation will be updated
* implement `Test`-corresponding macros like `@test_reference_broken`, `@test_reference_skip`

**Comment by johnnychen94** [comment]
mind to take a look at it? @oxinabox

File: src/core.jl [hunk 6]
```diff
@@ -76,10 +76,20 @@ end
 # all other functions should hit one of this eventually
 # Which handles the actual testing and user prompting
 
-function _test_reference(equiv, rendermode, file::File, actual::T) where T
+function test_reference(
+    file::File,
+    actual::T,
```
Comment (oxinabox) [inline src/core.jl]: Is this type constraint ever used?

File: src/core.jl [hunk 7]
```diff
@@ -76,10 +76,20 @@ end
...
+    rendermode::Union{RenderMode, Nothing}=nothing,
+    equiv::Union{Function, Nothing}=nothing;
```
Comment (oxinabox) [inline src/core.jl]: Rather than defaulting to `nothing` how do you feel about defaulting to `default_rendermode(file, actual)` ?

File: src/preprocess.jl [hunk 8]
```diff
@@ -0,0 +1,35 @@
+"""
+    preprocess(::File, actual; kw...) -> out
+
+Preprocess `actual` so that `out` is similar to reference file content.
+"""
+preprocess(::File, actual; kw...) = actual
```
Comment (oxinabox) [inline src/preprocess.jl]: Can we rename to to something more clear, like `convert_towards_ format`.
And maybe rather than taking the `File` just take the `Format` type

File: src/render.jl [hunk 9]
```diff
@@ -0,0 +1,19 @@
+"""
+    default_rendermode(::File, actual)
+
+Infer the most appropriate render mode according to type of reference file and `actual`.
+"""
+default_rendermode(::File, ::Any) = BeforeAfterLimited()
```
Comment (oxinabox) [inline src/render.jl]: Again, rather than taking the `File` taking the `Format` only seems better

File: src/core.jl [hunk 10]
```diff
@@ -100,7 +110,15 @@ function _test_reference(equiv, rendermode, file::File, actual::T) where T
         return nothing # skip current test case
     end
 
+    # file exists
+    actual = preprocess(file, actual; kw...)
```
Comment (oxinabox) [inline src/core.jl]: Can we not reused the `actual` variable?

Maybe before this line (in the function Sig) it should be `raw_actual` or something

File: src/core.jl (hunk 7 above)
Comment (johnnychen94) [reply src/core.jl]: Cool, I thought default arguments are evaluated only *once* in compilation time.

File: src/core.jl (hunk 6 above)
Comment (johnnychen94) [reply src/core.jl]: I added this to debug an error introduced by order changes. 
I think what I really need to do here is to swap the order of `rendermode` and `equiv` so that it's consistent to

```julia
test_reference(filename::AbstractString, actual; by = nothing, render = nothing, kw...)
```

**Comment by johnnychen94** [comment]
The last three commits do a trivial organization work. You can review the difference commit by commit.

File: src/render.jl (hunk 9 above)
Comment (johnnychen94) [reply src/render.jl]: done in https://github.com/Evizero/ReferenceTests.jl/pull/37/commits/8bb84a23c3776995332c7cf6091ce1bb1d53b5d9

File: src/core.jl (hunk 10 above)
Comment (johnnychen94) [reply src/core.jl]: done in https://github.com/Evizero/ReferenceTests.jl/pull/37/commits/8bb84a23c3776995332c7cf6091ce1bb1d53b5d9

File: src/core.jl (hunk 6 above)
Comment (johnnychen94) [reply src/core.jl]: fixed in https://github.com/Evizero/ReferenceTests.jl/pull/37/commits/8ea330db3e1383877ca9c144fd6da784cd48952d

File: src/core.jl (hunk 7 above)
Comment (johnnychen94) [reply src/core.jl]: It turns out this doesn't work, I've added a comment for this in https://github.com/Evizero/ReferenceTests.jl/pull/37/commits/8ea330db3e1383877ca9c144fd6da784cd48952d

File: src/preprocess.jl (hunk 8 above)
Comment (johnnychen94) [reply src/preprocess.jl]: renamed to `_convert` in https://github.com/Evizero/ReferenceTests.jl/pull/37/commits/8bb84a23c3776995332c7cf6091ce1bb1d53b5d9
